Tribler/community/multichain/block.py

```diff
@@ -0,0 +1,282 @@
+        if blk:
+            # TODO: if the block is valid upto this point (specifically the signature is valid), and an error is
+            # detected here, it is in fact fraud, since a PK/seq pair is double signed.
+            if blk.up != self.up:
+                err("Up does not match known block")
+            if blk.down != self.down:
+                err("Down does not match known block")
+            if blk.total_up != self.total_up:
+                err("Total up does not match known block")
+            if blk.total_down != self.total_down:
+                err("Total down does not match known block")
+            if blk.link_public_key != self.link_public_key:
+                err("Link public key does not match known block")
+            if blk.link_sequence_number != self.link_sequence_number:
+                err("Link sequence number does not match known block")
+            if blk.previous_hash != self.previous_hash:
+                err("Previous hash does not match known block")
+            if blk.signature != self.signature:
+                err("Signature does not match known block")
+
+        # Step 4: does the database have the linked block? If so do the values match up?
+        if link:
+            if self.up != link.down:
+                err("Up/down mismatch on linked block")
+            if self.down != link.up:
+                err("Down/up mismatch on linked block")
+
+        # Step 5: does the database have adjacent blocks?
+        if prev_blk:
+            if prev_blk.total_up + self.up > self.total_up:
+                err("Total up is lower than expected compared to the preceding block")
+            if prev_blk.total_down + self.down > self.total_down:
+                err("Total down is lower than expected compared to the preceding block")
+            if prev_blk.sequence_number == self.sequence_number - 1 and prev_blk.hash != self.previous_hash:
+                err("Previous hash is not equal to the hash id of the previous block")
+
+        if next_blk:
+            if self.total_up + next_blk.up > next_blk.total_up:
+                err("Total up is higher than expected compared to the next block")
+            if self.total_down + next_blk.down > next_blk.total_down:
+                err("Total down is higher than expected compared to the next block")
+            if next_blk.sequence_number == self.sequence_number + 1 and next_blk.previous_hash != self.hash:
+                err("Next hash is not equal to the hash id of the block")
+
+        # TODO: Detect fraud? But should it go here?
+        # Fraud detection might be easier if you inspect the database for all public key's in one go. (Or rather it
+        # might lead to a lot of double work if you do the fraud detection for each PK individually)
+        # Also if this method returns "invalid" for fraud cases, then those blocks cannot be kept in the database for
+        # future reference or as fraud proofs.
```

Comment:
I deliberately choose one level of invalidity. Considering how little I know about the exact implementation details of dispersy, even though we depend on it, I'm assuming future programmers on Tribler might be as clueless as me regarding the finer points of MultiChain blocks. However interpreting invalidity correctly is extremely important. Thus keeping the contract/API as simple as possible is a concern. With one level, you only need one check which you can't mess up.

This is also the reason why the validator returns messages about exactly what makes a block invalid. If there are places where you really want to distinguish about why it is different, you can use the validation strings for this. Perhaps we should make constants of these too (so code checking for specific conditions doesn't break if the actual string is updated).